### PR TITLE
Profile Configure Conditional Question Support

### PR DESCRIPTION
**Pull request by Cena-Studio**
# Overview

Along with the development of OpenFlight, the future `flight profile configure` may need to include more questions to provide users with a more comprehensive and customized service. Some of these questions will need to be determined based on the answer to the previous question, i.e., parent questions. Finally, the question library will form a tree-like branching structure because of those sub-questions. This PR aims to provide better structural support for such questions, allowing question designers to easily design and manage questions through YAML files. In addition, This PR also adds support for specifying a certain type of question that requires the answers to be only yes/y or no/n.

# Major Changes
- Conditional Question
   - Metadata: `type` is added as one of the question attributes. Currently, only `conditional` is a valid value for this parameter.
   - Metadata: The `questions` option can now also be written under a parent question to define its child questions. The child question itself can also have child questions.
   - Metadata: `where` is added as an attribute of child questions to specify when it should be presented based on the answer to the parent question.
   - UIX: A new form of question may be asked, which only allows the answer to be yes/y or no/n.
   - Implementation:  A new method, `Type#recursive_questions`, is added to recursively fetch all the questions and their child questions.
   - Implementation: `Configure#generate_prefill` is now changed to `Configure#generate_prefills` to generate prefills of all the questions in one method. `Configure#ask_questions` no longer creates separate threads for the prefill generation of each question.
   - Implementation: `Configure#collect_answers` method is added to replace the previous answers collecting algorithm in `Configure#ask_questions` to determine the type of the question (if it is a "conditional" question) and whether it needs to be presented.
   - Implementation: `Configure#use_cli_answers` is modified to fit the above functional changes. It is supported by a newly added method `Configure#required_cli_answers`.

# Examples
## Question Metadata YAML File Demo (Simplified)
```
questions:
  - id: question_1
    type: boolean
    questions:
      - id: question_1_1
        where: true
  - id: question_2
    questions:
      - id: question_2_1
        type: boolean
        where: 3n
        questions:
          - id: question_2_1_true
            where: true
          - id: question_2_1_false
            where: false
      - id: question_2_2
        where: 3n+1
      - id: question_2_3
        where: 3n+2
  - id: question_3
```
As can be seen from the above example, `question_1` is defined as a conditional question by the `type` parameter. `question_1_1` will only be raised when the answer to `question_1` is true as defined by the `where` parameter of `question_1_1`.
`question_2` is a relatively more complicated example that contains three levels of hierarchies. In that case, `question_2_1_false` will only be raised when the user answers "3n" to `question_2` and then answers "no" or "n" for `question_2_1`.
Please notice that the given example is simplified. For a real metadata.yaml, `id`, `name`, and `description` should be included to specify the attributes of the cluster. Besides, `env`, `text`, `default`, and `validation` should also be included as the required parameters for each question.

### CODE REVIEW

**Comment by VoxSecundus** [comment]
Please could you switch the target branch to `develop` and resolve the resulting merge conflict?

**Comment by ColonelPanicks** [comment]
I think flight profile is still struggling with true/false variables, see:

```bash
[flight@gateway1 ~]$ flight profile configure
Cluster type:  Openflight Slurm Multinode
Cluster name: my-cluster
NFS server (hostname or flight-hunter label): gateway1
SLURM server (hostname or flight-hunter label): gateway1
Setup Multi User Environment with IPA? No
Local user login: flight
Set local user password to: 0penfl1ght
IP or FQDN for Web Access: 10.151.15.151
IP Range of Compute Nodes: 10.100.0.0/16

[root@gateway1 profile]# cat var/answers/openflight-slurm-multinode.yaml
---
cluster_name: my-cluster
nfs_server: gateway1
slurm_server: gateway1
ipa_use: false
default_username: flight
default_password: 0penfl1ght
access_host: 10.151.15.151
compute_ip_range: 10.100.0.0/16

[flight@gateway1 ~]$ flight profile list
┌──────────┬──────────┬───────────┐
│ Node     │ Identity │ Status    │
├──────────┼──────────┼───────────┤
│ gateway1 │          │ available │
│ node01   │          │ available │
│ node02   │          │ available │
└──────────┴──────────┴───────────┘

[flight@gateway1 ~]$ flight profile apply gateway1 login
profile: The following config keys have not been set:
setup multi user environment with IPA?
Please run `profile configure`
```

My `metadata.yaml` for `openflight-slurm-multinode`

```yaml
---
id: openflight-slurm-multinode
name: 'Openflight Slurm Multinode'
description: 'A Multi Node Research Environment running SLURM & NFS'
questions:
  - id: cluster_name
    env: CLUSTERNAME
    text: 'Cluster name:'
    default: my-cluster
    validation:
      type: string
      format: "^[a-zA-Z0-9_\\-]+$"
      message: 'Invalid cluster name: %{value}. Must contain only alphanumeric characters,
      - and _.'
  - id: nfs_server
    env: NFS_SERVER
    text: 'NFS server (hostname or flight-hunter label):'
    default: 'login1'
    default_smart:
      - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
      - "hostname -s"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: slurm_server
    env: SLURM_SERVER
    text: 'SLURM server (hostname or flight-hunter label):'
    default: 'login1'
    default_smart:
      - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
      - "hostname -s"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: ipa_use
    env: IPA_USE
    text: 'Setup Multi User Environment with IPA?'
    default: no
    type: conditional
    validation:
      type: string
      format: "^(?i)(true|false)$"
      message: "Must be either true or false"
    questions:
      - id: ipa_server
        env: IPA_SERVER
        text: 'IPA server (short hostname or flight-hunter label):'
        default: 'infra01'
        where: true
        validation:
          type: string
          format: "^(\\S)+$"
          message: "Cannot be empty or contain whitespace"
      - id: ipa_domain
        env: IPA_DOMAIN
        text: 'IPA domain:'
        default: 'cluster.example.com'
        where: true
        default_smart:
          - 'hostname -d'
        validation:
          type: string
          format: "^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
          message: "Domain name must contain at least 2 elements separated by '.' (e.g. 'example.com')"
      - id: secure_admin_password
        env: SECURE_ADMIN_PASSWORD
        text: 'IPA Secure Admin Password:'
        default: "MySecurePassword"
        where: true
        validation:
          type: string
          format: "^(\\S)+$"
          message: "Cannot be empty or contain whitespace"
  - id: default_username
    env: DEFAULT_USERNAME
    text: 'Local user login:'
    default: flight
    default_smart:
      - "cut -d: -f1,3 /etc/passwd | egrep ':[0-9]{4}$' | cut -d: -f1 |head -1" # First local user with UID of 1000 or higher
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: default_password
    env: DEFAULT_PASSWORD
    text: 'Set local user password to:'
    default: 0penfl1ght
    validation:
      type: string
  - id: access_host
    env: ACCESS_HOST
    text: 'IP or FQDN for Web Access:'
    default_smart:
      - "curl -s --fail http://169.254.169.254/latest/meta-data/public-hostname |grep -v novalocal" # Public Hostname from AWS/OpenStack Metadata (excluding default OpenStack domain name)
      - "curl -s --fail http://169.254.169.254/latest/meta-data/public-ipv4" # IP from AWS/OpenStack Metadata
      - "curl -s --fail -H Metadata:true http://169.254.169.254/metadata/instance/network/interface/0/ipv4/ipAddress/0?api-version=2021-12-13 |sed 's/.*publicIpAddress\":\"//g;s/\".*//g'" # IP from Azure Metadata
      - "curl -s ifconfig.me" # External IP curl
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: compute_ip_range
    env: COMPUTE_IP_RANGE
    text: 'IP Range of Compute Nodes:'
    default: '10.10.0.0/16'
    default_smart:
      - "ipcalc $(/usr/sbin/ip addr show $(/usr/sbin/ip -o -4 route show to default |awk '{print $5}') |grep 'inet ' |sed 's/.*inet //g;s/ brd.*//g') |grep '^Network:' |awk '{print $2}'"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
```

**Comment by Cena-Studio** [comment]
> I think flight profile is still struggling with true/false variables, see:
> 
> ```shell
> [flight@gateway1 ~]$ flight profile configure
> Cluster type:  Openflight Slurm Multinode
> Cluster name: my-cluster
> NFS server (hostname or flight-hunter label): gateway1
> SLURM server (hostname or flight-hunter label): gateway1
> Setup Multi User Environment with IPA? No
> Local user login: flight
> Set local user password to: 0penfl1ght
> IP or FQDN for Web Access: 10.151.15.151
> IP Range of Compute Nodes: 10.100.0.0/16
> 
> [root@gateway1 profile]# cat var/answers/openflight-slurm-multinode.yaml
> ---
> cluster_name: my-cluster
> nfs_server: gateway1
> slurm_server: gateway1
> ipa_use: false
> default_username: flight
> default_password: 0penfl1ght
> access_host: 10.151.15.151
> compute_ip_range: 10.100.0.0/16
> 
> [flight@gateway1 ~]$ flight profile list
> ┌──────────┬──────────┬───────────┐
> │ Node     │ Identity │ Status    │
> ├──────────┼──────────┼───────────┤
> │ gateway1 │          │ available │
> │ node01   │          │ available │
> │ node02   │          │ available │
> └──────────┴──────────┴───────────┘
> 
> [flight@gateway1 ~]$ flight profile apply gateway1 login
> profile: The following config keys have not been set:
> setup multi user environment with IPA?
> Please run `profile configure`
> ```
> 
> My `metadata.yaml` for `openflight-slurm-multinode`
> 
> ```yaml
> ---
> id: openflight-slurm-multinode
> name: 'Openflight Slurm Multinode'
> description: 'A Multi Node Research Environment running SLURM & NFS'
> questions:
>   - id: cluster_name
>     env: CLUSTERNAME
>     text: 'Cluster name:'
>     default: my-cluster
>     validation:
>       type: string
>       format: "^[a-zA-Z0-9_\\-]+$"
>       message: 'Invalid cluster name: %{value}. Must contain only alphanumeric characters,
>       - and _.'
>   - id: nfs_server
>     env: NFS_SERVER
>     text: 'NFS server (hostname or flight-hunter label):'
>     default: 'login1'
>     default_smart:
>       - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
>       - "hostname -s"
>     validation:
>       type: string
>       format: "^(\\S)+$"
>       message: "Cannot be empty or contain whitespace"
>   - id: slurm_server
>     env: SLURM_SERVER
>     text: 'SLURM server (hostname or flight-hunter label):'
>     default: 'login1'
>     default_smart:
>       - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
>       - "hostname -s"
>     validation:
>       type: string
>       format: "^(\\S)+$"
>       message: "Cannot be empty or contain whitespace"
>   - id: ipa_use
>     env: IPA_USE
>     text: 'Setup Multi User Environment with IPA?'
>     default: no
>     type: conditional
>     validation:
>       type: string
>       format: "^(?i)(true|false)$"
>       message: "Must be either true or false"
>     questions:
>       - id: ipa_server
>         env: IPA_SERVER
>         text: 'IPA server (short hostname or flight-hunter label):'
>         default: 'infra01'
>         where: true
>         validation:
>           type: string
>           format: "^(\\S)+$"
>           message: "Cannot be empty or contain whitespace"
>       - id: ipa_domain
>         env: IPA_DOMAIN
>         text: 'IPA domain:'
>         default: 'cluster.example.com'
>         where: true
>         default_smart:
>           - 'hostname -d'
>         validation:
>           type: string
>           format: "^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
>           message: "Domain name must contain at least 2 elements separated by '.' (e.g. 'example.com')"
>       - id: secure_admin_password
>         env: SECURE_ADMIN_PASSWORD
>         text: 'IPA Secure Admin Password:'
>         default: "MySecurePassword"
>         where: true
>         validation:
>           type: string
>           format: "^(\\S)+$"
>           message: "Cannot be empty or contain whitespace"
>   - id: default_username
>     env: DEFAULT_USERNAME
>     text: 'Local user login:'
>     default: flight
>     default_smart:
>       - "cut -d: -f1,3 /etc/passwd | egrep ':[0-9]{4}$' | cut -d: -f1 |head -1" # First local user with UID of 1000 or higher
>     validation:
>       type: string
>       format: "^(\\S)+$"
>       message: "Cannot be empty or contain whitespace"
>   - id: default_password
>     env: DEFAULT_PASSWORD
>     text: 'Set local user password to:'
>     default: 0penfl1ght
>     validation:
>       type: string
>   - id: access_host
>     env: ACCESS_HOST
>     text: 'IP or FQDN for Web Access:'
>     default_smart:
>       - "curl -s --fail http://169.254.169.254/latest/meta-data/public-hostname |grep -v novalocal" # Public Hostname from AWS/OpenStack Metadata (excluding default OpenStack domain name)
>       - "curl -s --fail http://169.254.169.254/latest/meta-data/public-ipv4" # IP from AWS/OpenStack Metadata
>       - "curl -s --fail -H Metadata:true http://169.254.169.254/metadata/instance/network/interface/0/ipv4/ipAddress/0?api-version=2021-12-13 |sed 's/.*publicIpAddress\":\"//g;s/\".*//g'" # IP from Azure Metadata
>       - "curl -s ifconfig.me" # External IP curl
>     validation:
>       type: string
>       format: "^(\\S)+$"
>       message: "Cannot be empty or contain whitespace"
>   - id: compute_ip_range
>     env: COMPUTE_IP_RANGE
>     text: 'IP Range of Compute Nodes:'
>     default: '10.10.0.0/16'
>     default_smart:
>       - "ipcalc $(/usr/sbin/ip addr show $(/usr/sbin/ip -o -4 route show to default |awk '{print $5}') |grep 'inet ' |sed 's/.*inet //g;s/ brd.*//g') |grep '^Network:' |awk '{print $2}'"
>     validation:
>       type: string
>       format: "^(\\S)+$"
>       message: "Cannot be empty or contain whitespace"
> ```

Thank you for pointing this out. It was caused by the same reason as the previous one but in another method. I've fixed and successfully tried to apply a node with the given metadata. Btw plz remove `validation format` for the conditional type questions since the answer won't be a string and cannot be validated by a regex.

**Comment by ColonelPanicks** [comment]
> Btw plz remove validation format for the conditional type questions since the answer won't be a string and cannot be validated by a regex.

If I remove it I get the following error (as validation is _required_ by Flight Profile questions - not necessarily sure I agree that it should):
```
[flight@gateway1 ~]$ flight profile configure
Cluster type:  Openflight Slurm Multinode
Cluster name: condition1
NFS server (hostname or flight-hunter label): gateway1
SLURM server (hostname or flight-hunter label): gateway1
profile: undefined method `required' for nil:NilClass
```

**Comment by ColonelPanicks** [comment]
I think this is still having issues in the case of answers when the conditional is true:
```bash
[flight@gateway1 ~]$ flight profile configure
Cluster type:  Openflight Slurm Multinode
Cluster name: condition4
NFS server (hostname or flight-hunter label): gateway1
SLURM server (hostname or flight-hunter label): gateway1
Setup Multi User Environment with IPA? yes
IPA server (short hostname or flight-hunter label): node01
IPA domain: conditionals4.alces.network
IPA Secure Admin Password: MySecurePassword
Local user login: flight
Set local user password to: 0penfl1ght
IP or FQDN for Web Access: 10.151.15.58
IP Range of Compute Nodes: 10.100.0.0/16
profile: undefined method `match' for true:TrueClass
```

The metadata.yaml for `openflight-slurm-multinode`:
```yaml
---
id: openflight-slurm-multinode
name: 'Openflight Slurm Multinode'
description: 'A Multi Node Research Environment running SLURM & NFS'
questions:
  - id: cluster_name
    env: CLUSTERNAME
    text: 'Cluster name:'
    default: my-cluster
    validation:
      type: string
      format: "^[a-zA-Z0-9_\\-]+$"
      message: 'Invalid cluster name: %{value}. Must contain only alphanumeric characters,
      - and _.'
  - id: nfs_server
    env: NFS_SERVER
    text: 'NFS server (hostname or flight-hunter label):'
    default: 'login1'
    default_smart:
      - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
      - "hostname -s"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: slurm_server
    env: SLURM_SERVER
    text: 'SLURM server (hostname or flight-hunter label):'
    default: 'login1'
    default_smart:
      - "/opt/flight/bin/flight hunter list --plain |grep $(hostname -f) |awk '{print $5}'"
      - "hostname -s"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: ipa_use
    env: IPA_USE
    type: conditional
    text: 'Setup Multi User Environment with IPA?'
    default: 'false'
    validation:
      type: string
      format: "^(?i)(true|false)$"
      message: "Must be either true or false"
    questions:
      - id: ipa_server
        env: IPA_SERVER
        text: 'IPA server (short hostname or flight-hunter label):'
        default: 'infra01'
        where: true
        validation:
          type: string
          format: "^(\\S)+$"
          message: "Cannot be empty or contain whitespace"
      - id: ipa_domain
        env: IPA_DOMAIN
        text: 'IPA domain:'
        default: 'cluster.example.com'
        default_smart:
          - 'hostname -d'
        where: true
        validation:
          type: string
          format: "^(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
          message: "Domain name must contain at least 2 elements separated by '.' (e.g. 'example.com')"
      - id: secure_admin_password
        env: SECURE_ADMIN_PASSWORD
        text: 'IPA Secure Admin Password:'
        default: "MySecurePassword"
        where: true
        validation:
          type: string
          format: "^(\\S)+$"
          message: "Cannot be empty or contain whitespace"
  - id: default_username
    env: DEFAULT_USERNAME
    text: 'Local user login:'
    default: flight
    default_smart:
      - "cut -d: -f1,3 /etc/passwd | egrep ':[0-9]{4}$' | cut -d: -f1 |head -1" # First local user with UID of 1000 or higher
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: default_password
    env: DEFAULT_PASSWORD
    text: 'Set local user password to:'
    default: 0penfl1ght
    validation:
      type: string
  - id: access_host
    env: ACCESS_HOST
    text: 'IP or FQDN for Web Access:'
    default_smart:
      - "curl -s --fail http://169.254.169.254/latest/meta-data/public-hostname |grep -v novalocal" # Public Hostname from AWS/OpenStack Metadata (excluding default OpenStack domain name)
      - "curl -s --fail http://169.254.169.254/latest/meta-data/public-ipv4" # IP from AWS/OpenStack Metadata
      - "curl -s --fail -H Metadata:true http://169.254.169.254/metadata/instance/network/interface/0/ipv4/ipAddress/0?api-version=2021-12-13 |sed 's/.*publicIpAddress\":\"//g;s/\".*//g'" # IP from Azure Metadata
      - "curl -s ifconfig.me" # External IP curl
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
  - id: compute_ip_range
    env: COMPUTE_IP_RANGE
    text: 'IP Range of Compute Nodes:'
    default: '10.10.0.0/16'
    default_smart:
      - "ipcalc $(/usr/sbin/ip addr show $(/usr/sbin/ip -o -4 route show to default |awk '{print $5}') |grep 'inet ' |sed 's/.*inet //g;s/ brd.*//g') |grep '^Network:' |awk '{print $2}'"
    validation:
      type: string
      format: "^(\\S)+$"
      message: "Cannot be empty or contain whitespace"
```

**Comment by ColonelPanicks** [comment]
From IRL: The above is an issue caused by having `type` in the `validation` field for the conditional question, it causes profile to try and match it as a string, this is not an issue with the code

**Comment by ColonelPanicks** [comment]
This LGTM. Will put together modifications to flight-profile-types to support this.

**Comment by VoxSecundus** [comment]
@Cena-Studio Please could you rebase this on `develop` before I review it? :slightly_smiling_face:

**Review by VoxSecundus — CHANGES_REQUESTED** [review]
Really great work here. I don't remember if supporting deeper nesting of conditional questions was an original requirement, but I'm pleasantly surprised to see that it works. One rebase-related change to make, first.

**Comment by VoxSecundus** [comment]
I've also pushed a rebase to this branch:
- Squashing the numerous line-endings fix commits into one
- Dropping the commit that had rewound a lot of work to an earlier version of `develop`

**Comment by Cena-Studio** [comment]
An issue that the profile apply process didn't fetch the answer based on conditional questions has been fixed

File: lib/profile/commands/configure.rb [hunk 1]
```diff
@@ -43,41 +43,90 @@ def use_cli_answers
...
+          prefill = cluster_type.fetch_answer(question.id)
+          if question.default_smart && prefill.nil?
+            process = Flight::Subprocess::Local.new(
+              env: {},
+              logger: smart_log,
+              timeout: 5,
+            )
+            result = process.run(question.default_smart, nil)
+            output = result.stdout.chomp
+            if !result.success?
```
[the remark this answers is not in the document]
Comment (Cena-Studio) [reply lib/profile/commands/configure.rb]: This has been fixed

**Review by VoxSecundus — APPROVED** [review]
This LGTM. @ColonelPanicks I'm going to give you a chance to test this as it's had changes since you last looked at it; whether you do or don't, please post in this thread when you're ready to OK it.

**Pull request merged**